src/check.rs

```diff
@@ -60,17 +99,133 @@ pub fn is_available(url: &Url, ctx: &CheckContext) -> Result<(), CheckError> {
         }
     }
 }
+cached_key_result! {
+    CHECK_FILE: SizedCache<String, HashSet<String>> = SizedCache::with_size(100);
+    Key = { link.to_string() };
+    fn fragments_from(
+        link: &Link,
+        fetch_html: impl Fn() -> Result<String, CheckError>
+    ) -> Result<HashSet<String>, CheckError> = {
+        fetch_html().map(|html| parse_fragments(&html))
+    }
+}
+
+fn is_fragment_available(
+    link: &Link,
+    fragment: &str,
+    html: impl Fn() -> Result<String, CheckError>,
+) -> Result<(), CheckError> {
+    let fragments = fragments_from(link, html);
+
+    match fragments {
+        Ok(fragments) => {
+            if fragments.contains(fragment) {
+                Ok(())
```

Comment:
You can get rid of another level of indent with an early return.
```suggestion
            if fragments.contains(fragment) {
                return Ok(());
            }
```